src/harness/reference_models/pre_iap_filtering/pre_iap_util.py

```diff
@@ -0,0 +1,186 @@
+  assignment of that PAL ID is returned as the PPA frequency range.
+  Args:
+    ppa_record: A PPA record dictionary.
+    pal_records: List of PAL record dictionaries.
+  Returns:
+    Operating frequency range of PPA.
+  Raises:
+    Exception: If the PAL records not found for PPA or if the frequency range in all
+      PAL records are not same for PPA.
+  """
+  pal_frequencies = []
+  ppa_pal_ids = ppa_record['ppaInfo']['palId']
+  num_ppa_pal = len(ppa_pal_ids)
+  # Get all the frequencies from the PAL records whose ID is present in the given PPA record
+  for pal_record in pal_records:
+    for pal_id in range(0, num_ppa_pal):
+      if pal_record['palId'] == ppa_pal_ids[pal_id]:
+        pal_frequency = pal_record['channelAssignment']['primaryAssignment']
+        pal_frequencies.append(pal_frequency)
+        break
+    else:
+      # Raise an exception if no matching PAL ID is found
+      raise Exception('PAL record %s not found for PPA ID %s' %(pal_record['palId'], ppa_record['id']))
+
+  # Compare the frequencies of all the PAL records of the PPA
+  for pal_frequency in pal_frequencies:
+    if pal_frequency != pal_frequencies[0]:
+      # If the frequency of the PAL records are not matching raise an exception
+      raise Exception('The frequency range in all PAL records are not same for PPA ID %s' %ppa_record['id'])
+
+  # Return the first frequency range of the first matching PAL ID
+  return pal_frequencies[0]
+
+def getFssNeighboringCbsdsWithGrants(cbsds, fss_record, distance):
+  """Get the list of all CBSDs that lie within 150 KMs of FSS and have at least one grant.
+
+  Args:
+    cbsds :  List of CbsdData dictionaries as defined in the SAS-SAS specification.
+    fss_record: A FSS record dictionary.
+  Returns:
+    List of CBSDs lying within 150 KMs of FSS and having at least one grant.
+  """
+
+  neighboring_cbsds_with_grants = []
+  for cbsd in cbsds:
+    distance_km, _, _ = vincenty.GeodesicDistanceBearing(
+        fss_record['record']['deploymentParam'][0]['installationParam']['latitude'],
+        fss_record['record']['deploymentParam'][0]['installationParam']['longitude'],
+        cbsd['registrationRequest']['installationParam']['latitude'],
+        cbsd['registrationRequest']['installationParam']['longitude'])
+    # Get the list of cbsds that are within 150kms from the FSS entity
+    if distance_km <= distance and cbsd['grants']:
+        neighboring_cbsds_with_grants.append(cbsd)
+  return neighboring_cbsds_with_grants
+
+
+def getFssNeighboringGwbl(gwbl_records, fss_records): 
+  """Get the list of all fss that lie within 150 KMs of GWBL and operating below 3700 MHz
+
+  Args:
+    fss_records:  List of FSS records dictionary.
+    gwbl_records: List of GWBL record dictionary.
+  Returns:
+    List of FSSs lying within 150 KMs of GWBL and operating below 3700000000.
+  """
+
+  list_of_fss_neighboring_gwbl = []
+  for fss_record in fss_records:
+    for gwbl_record in gwbl_records:
```

Comment:
For efficiency, you could put the frequency check here:
```
if  (fss_record['record']['deploymentParam'][0]
     ['operationParam']['operationFrequencyRange']['highFrequency'] >= 3700000000):
   continue
```
Question: Is it the highfrequency we need to check or the low frequency? Please verify in the spec.

Reply:
Modified   as suggested.     As per the spec, high frequency has to be checked